pkg/provisioner/controller.go

```diff
@@ -337,6 +318,39 @@ func (c *obcController) handleProvisionClaim(key string, obc *v1alpha1.ObjectBuc
 		Parameters:        class.Parameters,
 	}
 
+	// Fill in basic necessary object bucket info
+	setBasicOBInfo := func() {
+		setObjectBucketName(ob, key)
+		ob.Spec.StorageClassName = obc.Spec.StorageClassName
+		ob.Spec.ReclaimPolicy = options.ReclaimPolicy
+		ob.SetLabels(c.provisionerLabels)
+	}
+
+	// Should an error be returned, attempt to clean up the object store and API servers by
+	// calling the appropriate provisioner method.  In cases where Provision() or Revoke()
+	// return an err, it's likely that the ob == nil, hindering cleanup.
+	defer func() {
```

Comment:
can this block be moved even further down in the func, say right before the secret is created?

Reply:
I think it **can**, but this will mean that if the implementation of `Provision` or `Grant` doesn't clean up fully, then there will not be a later `Delete`/`Revoke` call made. I would err on the side of keeping this where it is to allow for more "slop" in the `Provision`/`Grant` implementation, but I am open to moving this to just before secret creation and updating the `Provision` and `Grant` docs to mention that if there are any errors during `Provision`/`Grant`, the implementation **must** fully clean up all of its resources.

In responding here, my thoughts do raise another consideration that is out of scope for this PR. If I imagine a case where the operator is `kill -9`ed or there is a full crash of the operator, it's possible that the object bucket provisioning will have created a bucket and user for the bucket that get orphaned (they aren't recorded as created but also aren't cleaned up after creation). I wonder if it would be good to have a cleanup cycle on operator start where Rook and/or the provisioner would be able to find orphaned users/buckets/ObjectBuckets to delete them.

Reply:
In the past we thought about a cleanup cycle. I think it's probably a good idea.